### PR TITLE
add test of internal/db/rdb/mysql

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I added the test for `internal/db/rdb/mysql.go`
for creating test, there are some refactors or updates.
  1. add `dbr` package, which is wrapped current dbr package, for test with mock
  2. refactor for adopting single and multiple deleteMeta pattern.
    - https://github.com/vdaas/vald/blob/test/internal/db/rdb/mysql/internal/db/rdb/mysql/mysql.go#L352-L376
    - #725 

### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/db/rdb/mysql/dbr/select.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+type SelectStmt interface {
+	// From(table interface{}) SelectStmt
+	// Where(query interface{}, value ...interface{}) SelectStmt
+	// Limit(n uint64) SelectStmt
+	LoadContext(ctx context.Context, value interface{}) (int, error)
+}
+
+func (stmt *selectStmt) From(table interface{}) SelectStmt {
+	return stmt.SelectStmt.From(table)
```
Comment (kpango) [inline internal/db/rdb/mysql/dbr/select.go:36]:
```suggestion
	stmt.SelectStmt = stmt.SelectStmt.From(table)
	return  stmt
```

File: internal/db/rdb/mysql/dbr/insert.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	ExecContext(ctx context.Context) (sql.Result, error)
+	// Record(structValue interface{}) InsertStmt
+}
+
+func (stmt *insertStmt) Columns(column ...string) InsertStmt {
+	return stmt.InsertStmt.Columns(column...)
+}
+
+func (stmt *insertStmt) ExecContext(ctx context.Context) (sql.Result, error) {
+	return stmt.ExecContext(ctx)
```
Comment (hlts2) [inline internal/db/rdb/mysql/dbr/insert.go:40]:
```suggestion
	return stmt.InsertStmt.ExecContext(ctx)
```

File: internal/db/rdb/mysql/dbr/dbr.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+type (
+	Builder       = dbr.Builder
+	Connection    = dbr.Connection
+	EventReceiver = dbr.EventReceiver
+)
+
+type db struct{}
+
+type DBR interface {
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/dbr.go:30]: please define interface before the struct definition

File: internal/db/rdb/mysql/dbr/delete.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	"database/sql"
+
+	dbr "github.com/gocraft/dbr/v2"
+)
+
+type deleteStmt struct {
+	*dbr.DeleteStmt
+}
+
+type DeleteStmt interface {
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/delete.go:29]: same as https://github.com/vdaas/vald/pull/659/files#r481853965

File: internal/db/rdb/mysql/dbr/insert.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	"database/sql"
+
+	"github.com/gocraft/dbr"
+)
+
+type insertStmt struct {
+	*dbr.InsertStmt
+}
+
+type InsertStmt interface {
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/insert.go:29]: same as https://github.com/vdaas/vald/pull/659/files#r481853965

File: internal/db/rdb/mysql/mysql.go [hunk 6]
```diff
@@ -122,7 +125,7 @@ func (m *mySQLClient) Open(ctx context.Context) error {
 	conn.SetMaxIdleConns(m.maxIdleConns)
 	conn.SetMaxOpenConns(m.maxOpenConns)
 
-	m.session = conn.NewSession(nil)
+	m.session = dbr.NewSession(*conn, nil)
```
Comment (kevindiu) [inline internal/db/rdb/mysql/mysql.go:128]: can we put `NewSession()` to `m.dbr`?

Comment (vankichi) [reply internal/db/rdb/mysql/mysql.go:128]: Did you mean `m.session = m.dbr.NewSession(*conn, nil)` ?

Comment (vankichi) [reply internal/db/rdb/mysql/mysql.go:128]: @kevindiu

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go:128]: @vankichi 🤔🤔 maybe current design is better than `m.dbr.NewSession()` 🤔

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go:128]: but can we mock `dbr.Connection`?

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go:128]: if we can do this, we can follow the origin code instead of creating our custom function

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go:128]: I am not sure if we can do it:
change the return type of https://github.com/vdaas/vald/pull/659/files#diff-c05bae65105c786c32926dfcd653e9a4R23 from `*Connection` to `Connection` and define our own mock `Connection` interface

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go:128]: I think it depends on use case and if we can create it individually

Comment (vankichi) [reply internal/db/rdb/mysql/mysql.go]: @kevindiu 
during considering, in this situation, we do not have to this.
`NewSession()` use the Connection which will be created by `m.dbr.New()`, in other word, `NewSession` is not individual.

Comment (kevindiu) [reply internal/db/rdb/mysql/mysql.go]: @vankichi 
I think both fine. What I thought that following the original API `conn.NewSession()`. 
If we create the mock of the `Connection`,
https://github.com/vdaas/vald/pull/659/files#diff-c05bae65105c786c32926dfcd653e9a4R29
and change the 
https://github.com/vdaas/vald/pull/659/files#diff-c05bae65105c786c32926dfcd653e9a4R39 
to `func (*db) Open(driver, dsn string, log EventReceiver) (Connection, error) {`, 
we can mock everything.

If you think your implementation is good enough, I think it is ok

**Review by kevindiu — APPROVED** [review]
LGTM. but if you have any concern about this please let me know :)
https://github.com/vdaas/vald/pull/659#discussion_r482799383

File: internal/db/rdb/mysql/mysql.go [hunk 7]
```diff
@@ -323,7 +326,7 @@ func (m *mySQLClient) SetMetas(ctx context.Context, metas ...MetaVector) error {
 	return tx.Commit()
 }
 
-func deleteMetaWithTx(ctx context.Context, tx *dbr.Tx, uuid string) error {
+func deleteMetaWithTx(ctx context.Context, tx dbr.Tx, uuid string, dbr dbr.DBR) error {
```
Comment (kevindiu) [inline internal/db/rdb/mysql/mysql.go:329]:
```suggestion
func  (dbr dbr.DBR) deleteMetaWithTx(ctx context.Context, tx dbr.Tx, uuid string) error {
```

Comment (kevindiu) [inline internal/db/rdb/mysql/mysql.go:329]: I think we should use receiver to use `m.DBR` instead.
```suggestion
func (m *mySQLClient) deleteMetaWithTx(ctx context.Context, tx dbr.Tx, uuid string) error {
```

File: internal/db/rdb/mysql/mysql.go [hunk 8]
```diff
@@ -348,7 +351,7 @@ func (m *mySQLClient) DeleteMeta(ctx context.Context, uuid string) error {
 	}
 	defer tx.RollbackUnlessCommitted()
 
-	err = deleteMetaWithTx(ctx, tx, uuid)
+	err = deleteMetaWithTx(ctx, tx, uuid, m.dbr)
```
Comment (kevindiu) [inline internal/db/rdb/mysql/mysql.go:354]:
```suggestion
	err = m.deleteMetaWithTx(ctx, tx, uuid)
```

**Review by kevindiu — DISMISSED** [review]
LGTM

**Comment by vankichi** [comment]
@kevindiu @kpango @hlts2 
Sorry too late confirmation.
I confirmed whether Vald will work correct if this change is merged into master branch via k8s cluster.
As a result, I could get metadata from mysql (which is empty before deploy vald including this change).
It means that this change won't cause any problem.

**Review by hlts2 — DISMISSED** [review]
LGTM

**Review by kevindiu — DISMISSED** [review]
LGTM

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/db/rdb/mysql/dbr/dbr.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
+//
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/dbr.go:1]: can you add package doc?

**Review by hlts2 — COMMENTED** [review]
Looks good to me!

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/db/rdb/mysql/dbr/dbr.go (hunk 9 above)
Comment (vankichi) [reply internal/db/rdb/mysql/dbr/dbr.go:1]: @kevindiu @hlts2 
updated :)
https://github.com/vdaas/vald/pull/659/commits/d3b35ca4c6c9e35c84596db19f7a91a1c76273c2

File: internal/db/rdb/mysql/dbr/connection.go [hunk 10]
```diff
@@ -14,6 +14,8 @@
 // limitations under the License.
 //
 
+// Package dbr provides the dbr functions for handling with databases.
```
Comment (kevindiu) [inline internal/db/rdb/mysql/dbr/connection.go:17]: please create a file `doc.go` and write the package there.

Comment (vankichi) [reply internal/db/rdb/mysql/dbr/connection.go:17]: Why doc.go is needed in this case 🤔 ?
I'm sorry, but I'm not sure.

File: internal/db/rdb/mysql/get.go [hunk 11]
```diff
@@ -14,6 +14,7 @@
 // limitations under the License.
 //
 
+// Package mysql provides the functions for handling vald's data with MySQL.
```
Comment (kevindiu) [inline internal/db/rdb/mysql/get.go:17]: After writing the package doc in `doc.go`, I think we can remove this line

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by hlts2 — DISMISSED** [review]
LGTM

**Pull request merged**